src/marqo/utils.py

```diff
@@ -66,3 +71,68 @@ def convert_dict_to_url_params(d: dict) -> str:
     url_encoded = urllib.parse.quote_plus(as_str)
     return url_encoded
 
+
+def deprecate_parameters(*deprecated_params):
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            """ Checks if any of deprecated key arguments are passed to the function and issues a warning if so.
+            Check is performed by comparing the values of the deprecated parameters to their default values."""
+            # Get the default values of the function
+            default_values = func.__defaults__
+
+            # Get the parameter names
+            signature = inspect.signature(func)
+            default_parameter_names = [
+                param.name for param in signature.parameters.values() if param.default is not inspect.Parameter.empty
+            ]
+
+            # Create a dictionary mapping parameter names to their default values
+            defaults_dict = dict(zip(default_parameter_names, default_values))
+
+            # Check if any of the deprecated parameters have values different from their defaults
+            deprecated_params_with_non_default_values = [
+                param for param in deprecated_params if kwargs.get(param) != defaults_dict.get(param)
+            ]
+
+            if deprecated_params_with_non_default_values:
+                deprecated_params_string = ', '.join(deprecated_params_with_non_default_values)
+                warning_msg = \
+                    f"The parameter(s) {deprecated_params_string} are deprecated. " \
+                    f"Please refer to the documentation " \
+                    f"https://docs.marqo.ai/1.3.0/Using-Marqo-Cloud/indexes/#create-index " \
+                    f"for updated parameters names. These parameters will be removed in Marqo 2.0.0."
+
+                # Issue a deprecation warning
+                mq_logger.warn(warning_msg)
+
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def is_key_arguments_default(current_frame, func, arguments_to_ignore=None):
+    """ Checks if all key arguments of the function are default values.
+    Ignores arguments specified in arguments_to_ignore list."""
+    passed_and_default = zip(current_frame.f_locals.items(), inspect.signature(func).parameters.values())
+    for passed_parameter, default_parameter in passed_and_default:
+        if default_parameter.default is not inspect.Parameter.empty:
+            if arguments_to_ignore and passed_parameter[0] in arguments_to_ignore:
+                continue
+            if passed_parameter[1] != default_parameter.default:
+                return False
+    return True
+
+
+def use_one_of_two_arguments(func, value_name, value, deprecated_value, deprecated_value_name):
+    """ Checks if both value and deprecated_value are passed to the function and returns only one of them,
+    prioritising non-deprecated one if so. Raises error if both of them are specified"""
```

Comment:
So the possible failure scenario is this:

```
my_func(new_param = 1, deprecated_param = 2):
      do_something() 

# user's code: 
my_func(new_param = 1, deprecated_param = 3)
# deprecated_param = 3 will be used instead of the expected new_param = 1
```